src/Transition.js

```diff
@@ -222,14 +207,19 @@ class Transition extends React.Component {
   }
 
   performEnter(mounting) {
-    const { enter } = this.props
+    const { enter, timeout } = this.props
     const appearing = this.context ? this.context.isMounting : mounting
     const [maybeNode, maybeAppearing] = this.props.nodeRef
       ? [appearing]
       : [ReactDOM.findDOMNode(this), appearing]
 
-    const timeouts = this.getTimeouts()
-    const enterTimeout = appearing ? timeouts.appear : timeouts.enter
+    const enterTimeout = typeof timeout !== 'number' && timeout ? timeout.enter : timeout
+    const appearTimeout = (typeof timeout !== 'number' && timeout
+      ? timeout.appear
+      : timeout
+    ) || enterTimeout // TODO: remove `enterTimeout` fallback for next major
+    const enterTransitionTimeout = appearing ? appearTimeout : enterTimeout
```

Comment:
**issue:** can simplify:

```suggestion
    let enterTimeout;
    if (typeof timeout !== 'number' && timeout) {
      // TODO: Remove appear-to-enter fallback for next major.
      enterTimeout = appearing && timeout.appear !== undefined
        ? timeout.appear
        : timeout.enter;
    } else {
      enterTimeout = timeout;
    }
```

note that, unlike in the previous version, your implementation would fall back from a falsy (but defined) `appear` to `enter`